src/package-linker.js

```diff
@@ -398,7 +398,7 @@ export default class PackageLinker {
       await promise.queue(
         topLevelDependencies,
         async ([dest, pkg]) => {
-          if (pkg.bin && Object.keys(pkg.bin).length) {
+          if (pkg._reference && pkg._reference.location && pkg.bin && Object.keys(pkg.bin).length) {
```

Comment:
I did something very similar here: https://github.com/yarnpkg/yarn/pull/4129/files#diff-306a2caea89d886b11134fb770d1ebf6R92

May be we need a wider check for this?